src/modules/utils/motordrivercontrol/drivers/TMC220X/TMC220X.cpp

```diff
@@ -0,0 +1,1870 @@
+    //set user defined amplitude offset
+    pwmconf_register_value |= (ofs << TMC220X_PWMCONF_PWM_OFS_SHIFT);
+
+    //if started we directly send it to the motor
+    if (started) {
+        transceive220X(TMC220X_WRITE|TMC220X_PWMCONF_REGISTER,pwmconf_register_value);
+    }
+}
+
+void TMC220X::setHolddelay(uint8_t value)
+{
+    //perform some sanity checks
+    if (value > 15) {
+        value = 15;
+    }
+
+    //delete the old value
+    this->ihold_irun_register_value &= ~(TMC220X_IHOLD_IRUN_IHOLDDELAY);
+
+    //save the new value
+    this->ihold_irun_register_value = value << TMC220X_IHOLD_IRUN_IHOLDDELAY_SHIFT;
+
+    //if started we directly send it to the motor
+    if (started) {
+        transceive220X(TMC220X_WRITE|TMC220X_IHOLD_IRUN_REGISTER,ihold_irun_register_value);
+    }
+}
+
+void TMC220X::setPowerDowndelay(uint8_t value)
+{
+    //save the delay to the register variable
+    this->tpowerdown_register_value = value;
+
+    //if started we directly send it to the motor
+    if (started) {
+        transceive220X(TMC220X_WRITE|TMC220X_TPOWERDOWN_REGISTER,tpowerdown_register_value);
+    }
+}
+
+void TMC220X::setStealthChopthreshold(uint32_t threshold)
+{
+    //perform some sanity checks
+    if (threshold >= (1 << 20)) {
+        threshold = (1 << 20) - 1;
+    }
+
+    //save the threshold value
+    this->tpwmthrs_register_value = threshold;
+
+    //if started we directly send it to the motor
+    if (started) {
+        transceive220X(TMC220X_WRITE|TMC220X_TPWMTHRS_REGISTER,tpwmthrs_register_value);
+    }
+}
+
+void TMC220X::setCoolThreshold(uint32_t threshold)
+{
+    //perform some sanity checks
+    if (threshold >= (1 << 20)) {
+        threshold = (1 << 20) - 1;
+    }
+
+    //save the threshold value
+    this->tcoolthrs_register_value = threshold;
+
+    //if started we directly send it to the motor
+    if (started && chip_type == StepstickParameters::CHIP_TYPE::TMC2209) {
+        transceive220X(TMC220X_WRITE|TMC220X_TCOOLTHRS_REGISTER,tcoolthrs_register_value);
+    }
+}
+
+void TMC220X::setStallguardThreshold(uint32_t threshold)
+{
+    //perform some sanity checks
+    if (threshold >= (1 << 8)) {
+        threshold = (1 << 8) - 1;
+    }
+
+    //save the threshold value
+    this->sgthrs_register_value = threshold;
+
+    //if started we directly send it to the motor
+    if (started && chip_type == StepstickParameters::CHIP_TYPE::TMC2209) {
+        transceive220X(TMC220X_WRITE|TMC220X_TCOOLTHRS_REGISTER,tcoolthrs_register_value);
+    }
+}
+
+uint8_t TMC220X::getStallguardResult(void)
+{
+    //if we don't yet started there cannot be a current value
+    if (!started || chip_type != StepstickParameters::CHIP_TYPE::TMC2209) {
+        return 0;
+    }
+    // Bits 9 and 0 will always show 0. 
+    // Scaling to 10 bit is for compatibility to StallGuard2.
+    unsigned long result = readRegister(TMC220X_SG_RESULT_REGISTER);
+
+    return (uint8_t)result;
```

Comment:
Dropping bit 8 on the floor.  Better to just return an int so the values make sense when  comparing to sgthreshold.